File: graphs & tries/797.js
class Solution{
    dfs(graph, u, target, res, temp){
        temp.push(u)
        if(u === target){
            res.push([...temp])
        }else{
            for(let v of graph[u]){
                this.dfs(graph, v, target, res, temp)
            }
        }
        temp.pop()
    }
    allPathsSourceTarget(graph){
        const n = graph.length
        const source = 0
        const res = []
        const temp = []
        const target = n - 1
        this.dfs(graph, source, target, res, temp)
        return res
    }
}

File: graphs & tries/797.py
# Amazon, Microsoft
# Directed Cyclic Graph(DAG)
# Tc = O(n)
# Sc = O(n)
class Solution:
    def dfs(self, graph, u, target, result, temp):
        temp.append(u)
        if u == target:
            result.append(temp[:])
        else:
            for v in graph[u]:
                self.dfs(graph, v, target, result, temp)
        temp.pop()

    def allPathSourceTarget(self, graph):
        n = len(graph)
        source = 0
        target = n - 1
        result = []
        temp = []
        self.dfs(graph, source, target, result, temp)
        return result